### ISSUE TITLE
Graphical glitches when large units move across the map

### ISSUE THREAD

**Issue by Toranks** [wesnoth/wesnoth#7626]
### Game and System Information

The Battle for Wesnoth version 1.17.16 x86_64
Running on Microsoft Windows 7 Service Pack 1 (6.1.7601) x86_64
Distribution channel: Steam

Libraries
=========

Boost:             1.81
Lua:               5.4.4
OpenSSL/libcrypto: 3.0.0h-dev (runtime 3.0.0h-dev)
Cairo:             1.17.8 (runtime 1.17.8)
Pango:             1.50.14 (runtime 1.50.14)
SDL:               2.26.3 (runtime 2.26.3)
SDL_image:         2.6.3 (runtime 2.6.3)
SDL_mixer:         2.6.3 (runtime 2.6.3)

Features
========

Lua console completion:       yes
Win32 notifications back end: yes

Current video settings
======================

SDL video drivers:        [windows] offscreen dummy
Window size:              1920x1080
Game canvas size:         1920x1080
Final render target size: 1920x1080
Screen refresh rate:      60
Screen dpi:               96.00x96.00
Renderer:                 direct3d (hw)
Maximum texture size:     16384x16384
VSync:                    on

### Description of the bug

When a very large unit moves across the map, each time the map is scrolled to track its movement, it leaves behind a piece of afterimage. This image disappears if I do zoom in/out during the movement or when the movement is finished.

https://github.com/wesnoth/wesnoth/assets/40789364/7639eb76-2aeb-4d3a-a39e-de70bed3f655

**Comment by ProditorMagnus** [comment]
You realize the expected action for this issue is to add validation that unit images must fit in 72x72.

**Comment by Toranks** [comment]
That would break a lot of mods, but in 1.16 it doesn't happen, so it's a problem with the new graphics engine.

**Comment by Slayer95** [comment]
> That would break a lot of mods, but in 1.16 it doesn't happen, so it's a problem with the new graphics engine.

I have seen that happen in 1.16 with To Lands Unknown under debug mode, too.

**Comment by Toranks** [comment]
This is in 1.16. What you say must be a different problem, due to the gigantic sprites it uses, some of which occupy half the screen.

https://github.com/wesnoth/wesnoth/assets/40789364/1ac50460-b00a-4e9f-9b54-572358bbb495

**Comment by CelticMinstrel** [comment]
> You realize the expected action for this issue is to add validation that unit images must fit in 72x72.

This sounds like a pretty unreasonable course of action. It's clear the system doesn't rely on it, and there are many units that rely on the restriction not existing.

The proper approach would be to not assume the unit fits in 72x72 and instead use the real bounding box. An alternate acceptable approach would be to allow explicitly defining the bounding box in WML.

**Comment by doofus-01** [comment]
I couldn't reproduce just now, and I can't see these videos, but whatever is going on, the suggestion of a new 72px requirement has got to be a strange joke.

I assume the issue is out-of-hex, not out-of-72px-box?  Sometimes there are issues with movement because of how standing animations are defined (and I might be guilty of adding such a glitched animation to core), which units are having a problem, specifically?

**Comment by Toranks** [comment]
`which units are having a problem, specifically?`

I will use this thread to report other units with the same error, in case it helps to discover the pattern of the failure

**Comment by CelticMinstrel** [comment]
> I assume the issue is out-of-hex, not out-of-72px-box?

That's unclear. It could be either as far as I know.

> I couldn't reproduce just now, and I can't see these videos

In the video, what happens is a Fire Dragon recolour is moving across a large map. Every time the map scrolls, the dragon leaves artifacts behind on the hex it occupied at that time.

**Comment by doofus-01** [comment]
> a Fire Dragon recolour

OK, so it might be how the animations are defined, but the idea seems less likely now.  I don't think the mainline fire dragon has a configuration issue, but if it also has the visual bug, we could try giving it a manually defined movement animation.

**Comment by Toranks** [comment]
It also happens with the mainline dragon, but not with the skeletal dragon.

https://github.com/wesnoth/wesnoth/assets/40789364/1f164f49-0914-4b51-a1aa-c8907bccefc3

**Comment by doofus-01** [comment]
> It also happens with the mainline dragon, but not with the skeletal dragon.

Hrm, this doesn't support the animation config theory either, I guess that idea was a red herring.  I'm not sure why it would affect the one but not the other, maybe the skeletal bits are just harder to see?  There's nothing obviously different in the metadata of the two images, unless resolution units matter somehow (cm vs inch)?

> Curiously, the LOTI dragon has animations of flying with its wings moving, but the mainline one does not. As a parallel question, couldn't you take advantage of those animations?

I just checked out LotI from 1.16, and see one frame.  There is nothing to take advantage of.

**Comment by Wedge009** [comment]
I confirmed this just now using the fire dragon. Having the unit travel far enough to cause a screen scroll seems to be a key part of replicating the issue.

**Comment by Wedge009** [comment]
I'm not sure what's going on specifically in that scenario Toranks is using. I replicated the issue using the attached save game: [7626.gz](https://github.com/wesnoth/wesnoth/files/11549544/7626.gz)

Take the fire dragon in the east and move to (1, 16). It leaves behind the top edges of its wings even on 1.16.9. Could be a slightly different issue but to me it looks the same as what Toranks shows in the original video capture.

**Comment by Toranks** [comment]
I can't replicate the problem with your savegame on 1.16.9. Also, your savegame says to be saved at 1.17.17dev+

**Comment by Wedge009** [comment]
Yes, but you can still load the game anyway. Here's what I get in 1.16.9 release on Linux - I also replicated on same version on Windows.

![image](https://github.com/wesnoth/wesnoth/assets/13679322/a77560ad-0d5e-4272-9ae1-4c1d2e947a0a)

Perhaps it is dependent on screen resolution - as I said above, the screen scroll seems central to the issue. I am running in 1920x1200 and 1600x900 windows.

**Comment by Wedge009** [comment]
Perhaps not related because I don't think a screen scroll was involved but I see this at the end of the Ancestor scenario in Winds of Fate, Wesnoth 1.18.2.

![image](https://github.com/wesnoth/wesnoth/assets/13679322/12484611-d62f-46bb-a3c5-0a00b758b602)

**Comment by mentos987** [comment]
I think this is something that should be fixed in unit_frame::get_overlaped_hex() in src/units/frame.cpp .

That function looks at the current location of an image and sees what hexes it overlaps with. And uses that to tell what hexes needs redrawing. The bug is that it misses some hex invalidations when you are scrolling at the same time. 

The solution is probably to check for current scrolling and add an offset for it in the detection area. Same as I recently did with offset in https://github.com/wesnoth/wesnoth/pull/10530/files#diff-e0c75959e8b77218263e7044774001287b3e1618e6d8b2462fe321ccf6b17377 for parallax unit image behaviour.

**Comment by mentos987** [comment]
Solved in: https://github.com/wesnoth/wesnoth/pull/10609
